subprojects/core/src/main/java/org/gradle/api/internal/file/copy/FileCopyAction.java

```diff
@@ -43,10 +45,20 @@ public WorkResult execute(CopyActionProcessingStream stream) {
         @Override
         public void processFile(FileCopyDetailsInternal details) {
             File target = fileResolver.resolve(details.getRelativePath().getPathString());
+            renameIfCaseChanged(target);
             boolean copied = details.copyTo(target);
             if (copied) {
                 didWork = true;
             }
         }
+
+        private void renameIfCaseChanged(File target) {
+            if (target.exists()) {
+                File canonicalizedTarget = FileUtils.canonicalize(target);
```

Comment:
should I attempt a rename instantly, if it exists, or only if the nanonical is different?